File: frontend/src/types/pipeline.ts
export interface PipelineStage {
  id: string
  name: string
  order_no: number
}

export interface InterviewRound {
  id: string
  interview_round: number
  status: string
  scheduled_datetime: string | null
  mode: string | null
  interviewer_name: string | null
}

export interface PipelineCard {
  id: string
  candidate_id: string
  candidate_name: string
  candidate_email: string
  candidate_phone: string | null
  current_designation: string | null
  current_company: string | null
  total_experience_years: number | null
  job_requirement_id: string
  job_title: string
  client_id: string
  client_name: string
  assigned_recruiter_id: string | null
  assigned_recruiter_name: string | null
  job_assignee_user_id: string | null
  job_assignee_name: string | null
  current_stage_id: string | null
  current_stage: string | null
  status: string
  owner_status: string
  applied_date: string | null
  submitted_at: string | null
  remarks: string | null
  interview_round: number | null
  interview_scheduled_at: string | null
  interview_mode: string | null
  interview_status: string | null
  interviewer_name: string | null
  interviews: InterviewRound[]
  offer_ctc: number | null
  offer_date: string | null
  offer_status: string | null
  offer_joining_date: string | null
  joined_date: string | null
}

export interface PipelineBoard {
  stages: PipelineStage[]
  cards: PipelineCard[]
}

export interface StageHistoryItem {
  id: string
  stage_id: string
  stage_name: string
  moved_by: string
  moved_by_name: string | null
  remarks: string | null
  created_at: string
}

export interface CandidateApplicationHistory {
  application_id: string
  job_requirement_id: string
  job_title: string
  client_id: string
  client_name: string
  current_stage: string | null
  owner_status: string
  status: string
  submitted_at: string | null
  history: StageHistoryItem[]
}

export interface CandidateStageHistory {
  candidate_id: string
  applications: CandidateApplicationHistory[]
}

export interface PipelineMovePayload {
  stage_name: string
  remarks?: string | null
  offered_ctc?: number | null
  joining_date?: string | null
  joined_date?: string | null
  revenue_generated?: number | null
  interview_scheduled_at?: string | null
  interviewer_name?: string | null
  interview_mode?: string | null
}

export type InterviewAction = 'cancel' | 'change_date' | 'no_show' | 'reschedule'

export interface InterviewActionPayload {
  action: InterviewAction
  interview_scheduled_at?: string | null
  interviewer_name?: string | null
  interview_mode?: string | null
  remarks?: string | null
}

export interface ShareApprovedPayload {
  to_emails?: string[]
  subject?: string
  message?: string
  application_ids?: string[]
}

export interface ShareApprovedResult {
  sent_to: string[]
  subject: string
  attachment: string | null
  candidate_count: number
}

/** Main hiring path; stage moves mirror the backend's immediate-next rules. */
export const FORWARD_ORDER = [
  'Applied',
  'Shortlisted',
  'Interview',
  'Offer',
  'Joined',
] as const

export function nextForwardStage(currentStage: string): string | null {
  const index = FORWARD_ORDER.indexOf(currentStage as (typeof FORWARD_ORDER)[number])
  return index >= 0 ? (FORWARD_ORDER[index + 1] ?? null) : null
}

export function resumeStageFromHistory(history: StageHistoryItem[]): string | null {
  for (const item of [...history].reverse()) {
    if (FORWARD_ORDER.includes(item.stage_name as (typeof FORWARD_ORDER)[number])) {
      return item.stage_name
    }
  }
  return null
}

export function allowedStageTargets(currentStage: string, resumeStage?: string | null): string[] {
  if (currentStage === 'Joined' || currentStage === 'Rejected') return []
  if (currentStage === 'On Hold') {
    return [...(resumeStage ? [resumeStage] : []), 'Rejected']
  }
  if (currentStage === 'Interview') {
    return ['Interview', 'Offer', 'On Hold', 'Rejected']
  }
  const nextStage = nextForwardStage(currentStage)
  return [...(nextStage ? [nextStage] : []), 'On Hold', 'Rejected']
}

/** @deprecated use allowedStageTargets */
export const PIPELINE_TRANSITIONS: Record<string, string[]> = {
  Applied: allowedStageTargets('Applied'),
  Shortlisted: allowedStageTargets('Shortlisted'),
  Interview: allowedStageTargets('Interview'),
  Offer: allowedStageTargets('Offer'),
  Joined: [],
  'On Hold': allowedStageTargets('On Hold'),
  Rejected: [],
}

File: backend/app/api/schemas/pipeline.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field


class PipelineStageResponse(BaseModel):
    id: uuid.UUID
    name: str
    order_no: int


class InterviewRoundResponse(BaseModel):
    id: uuid.UUID
    interview_round: int
    status: str
    scheduled_datetime: datetime | None = None
    mode: str | None = None
    interviewer_name: str | None = None


class PipelineCardResponse(BaseModel):
    id: uuid.UUID
    candidate_id: uuid.UUID
    candidate_name: str
    candidate_email: str
    candidate_phone: str | None = None
    current_designation: str | None = None
    current_company: str | None = None
    total_experience_years: float | None = None
    job_requirement_id: uuid.UUID
    job_title: str
    client_id: uuid.UUID
    client_name: str
    assigned_recruiter_id: uuid.UUID | None = None
    assigned_recruiter_name: str | None = None
    job_assignee_user_id: uuid.UUID | None = None
    job_assignee_name: str | None = None
    current_stage_id: uuid.UUID | None = None
    current_stage: str | None = None
    status: str
    owner_status: str
    in_pipeline: bool = False
    applied_date: date | None = None
    submitted_at: datetime | None = None
    remarks: str | None = None
    interview_round: int | None = None
    interview_scheduled_at: datetime | None = None
    interview_mode: str | None = None
    interview_status: str | None = None
    interviewer_name: str | None = None
    interviews: list[InterviewRoundResponse] = []
    offer_ctc: Decimal | None = None
    offer_date: date | None = None
    offer_status: str | None = None
    offer_joining_date: date | None = None
    joined_date: date | None = None


class PipelineBoardResponse(BaseModel):
    stages: list[PipelineStageResponse]
    cards: list[PipelineCardResponse]


class StageHistoryItem(BaseModel):
    id: uuid.UUID
    stage_id: uuid.UUID
    stage_name: str
    moved_by: uuid.UUID
    moved_by_name: str | None = None
    remarks: str | None = None
    created_at: datetime


class CandidateApplicationHistory(BaseModel):
    application_id: uuid.UUID
    job_requirement_id: uuid.UUID
    job_title: str
    client_id: uuid.UUID
    client_name: str
    current_stage: str | None = None
    owner_status: str
    status: str
    submitted_at: datetime | None = None
    history: list[StageHistoryItem]


class CandidateStageHistoryResponse(BaseModel):
    candidate_id: uuid.UUID
    applications: list[CandidateApplicationHistory]


class PipelineMoveRequest(BaseModel):
    stage_name: str
    remarks: str | None = None
    offered_ctc: Decimal | None = None
    joining_date: date | None = None
    joined_date: date | None = None
    revenue_generated: Decimal | None = None
    interview_scheduled_at: datetime | None = None
    interviewer_name: str | None = None
    interview_mode: str | None = None


class InterviewActionRequest(BaseModel):
    action: Literal["cancel", "change_date", "no_show", "reschedule"]
    interview_scheduled_at: datetime | None = None
    interviewer_name: str | None = None
    interview_mode: str | None = None
    remarks: str | None = None


class ShortlistRequest(BaseModel):
    application_ids: list[uuid.UUID] = Field(min_length=1)
    remarks: str | None = "Shortlisted by client"


class ShareApprovedRequest(BaseModel):
    to_emails: list[str] | None = None
    subject: str | None = None
    message: str | None = None
    application_ids: list[uuid.UUID] | None = None


class ShareApprovedResponse(BaseModel):
    sent_to: list[str]
    subject: str
    attachment: str | None = None
    candidate_count: int


class EmailStatusResponse(BaseModel):
    configured: bool
